dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/lb/TokenAwareSelection.java

```diff
@@ -32,77 +37,88 @@
 import com.netflix.dyno.connectionpool.impl.utils.CollectionUtils.Transform;
 
 /**
- * Concrete implementation of the {@link HostSelectionStrategy} interface using the TOKEN AWARE algorithm.
- * Note that this component needs to be aware of the dynomite ring topology to be able to 
- * successfully map to the correct token owner for any key of an {@link Operation}
+ * Concrete implementation of the {@link HostSelectionStrategy} interface using
+ * the TOKEN AWARE algorithm. Note that this component needs to be aware of the
+ * dynomite ring topology to be able to successfully map to the correct token
+ * owner for any key of an {@link Operation}
  * 
  * @author poberai
  *
  * @param <CL>
  */
 public class TokenAwareSelection<CL> implements HostSelectionStrategy<CL> {
+    
+    private final BinarySearchTokenMapper tokenMapper;
+
+    private final ConcurrentHashMap<Long, HostConnectionPool<CL>> tokenPools = new ConcurrentHashMap<Long, HostConnectionPool<CL>>();
+
+    public TokenAwareSelection() {
+
+        this.tokenMapper = new BinarySearchTokenMapper(new Murmur1HashPartitioner());
+    }
+
+    @Override
+    public void initWithHosts(Map<HostToken, HostConnectionPool<CL>> hPools) {
+
+        tokenPools.putAll(CollectionUtils.transformMapKeys(hPools, new Transform<HostToken, Long>() {
+
+            @Override
+            public Long get(HostToken x) {
+                return x.getToken();
+            }
+
+        }));
+
+        this.tokenMapper.initSearchMecahnism(hPools.keySet());
+    }
+
+    /**
+     * If a hashtag is provided by Dynomite then we use that to craete the key to hash.
+     */
+    @Override
+    public HostConnectionPool<CL> getPoolForOperation(BaseOperation<CL, ?> op, String hashtag) throws NoAvailableHostsException {
+
+        String key = op.getKey();
+
+        HostToken hToken = null;
+        if (hashtag == null || hashtag.isEmpty()) {            
+            hToken = this.getTokenForKey(key);
+        } else {  
+            String hashValue = StringUtils.substringBetween(key,Character.toString(hashtag.charAt(0)), Character.toString(hashtag.charAt(1)));
```

Comment:
To me, it seems that for Dynomite (https://github.com/Netflix/dynomite/blob/v0.5.9/src/dyn_message.c#L710) the key will include the hashtag. So for example, for a key ```foo{bar}foo``` and hashtag ```{}```, Dynomite will say the key is ```{bar}``` whereas your code will say the key is ```bar```.
Can you please check if I am correct?

Reply:
If the hashtag is defined as `{}` why would Dynomite say the key is `{bar}`? This is incorrect.

Reply:
correct. My bad!